### PR TITLE
stream: allocate enough space for select() hack

### PR DESCRIPTION
**Pull request by indutny**
`fd_set`s are way too small for `select()` hack when stream's fd is
bigger than 1023. Make `fd_set`s a part of `uv__stream_select_t`
structure.

Should be backported to all branches up to (and including) v0.10

### CODE REVIEW

**Comment by indutny** [comment]
cc @saghul

File: src/unix/stream.c [hunk 1]
```diff
@@ -318,26 +341,36 @@ int uv__stream_try_select(uv_stream_t* stream, int* fd) {
...
+failed_async_sem_init:
   uv_sem_destroy(&s->close_sem);
-fatal1:
+
+failed_close_sem_init:
+  uv__close(fds[0]);
+  uv__close(fds[1]);
   uv_close((uv_handle_t*) &s->async, uv__stream_osx_cb_close);
-  return -errno;
+  return err;
```
Comment (saghul) [inline src/unix/stream.c:362]: 's' needs to be freed here as well.

File: src/unix/stream.c [hunk 2]
```diff
@@ -318,26 +341,36 @@ int uv__stream_try_select(uv_stream_t* stream, int* fd) {
...
+
+failed_async_init:
+  free(s);
```
Comment (saghul) [inline src/unix/stream.c:365]: s is already bein freed before you jump here.

File: src/unix/stream.c [hunk 3]
```diff
@@ -283,32 +288,50 @@ int uv__stream_try_select(uv_stream_t* stream, int* fd) {
...
+    max_fd = fds[1];
+
+  sread_sz = (max_fd + NBBY) / NBBY;
+  swrite_sz = sread_sz;
+
   /* At this point we definitely know that this fd won't work with kqueue */
-  s = malloc(sizeof(*s));
-  if (s == NULL)
-    return -ENOMEM;
+  s = malloc(sizeof(*s) + sread_sz + swrite_sz);
```
Comment (saghul) [inline src/unix/stream.c:306]: Mode the comment above this line to the right place, that is, before creating the socketpair.

**Comment by saghul** [comment]
Left some comments. As for the commit log, shouldn't it be "darwin:"? This hack doesn't really apply to other platforms.

Test passes here, so I think we got this :-)

**Comment by saghul** [comment]
Note: we need to merge this into v1.x, not master.

File: src/unix/stream.c (hunk 1 above)
Comment (indutny) [inline src/unix/stream.c:362]: It is freed in `uv__stream_osx_cb_close`, right?

Comment (saghul) [inline src/unix/stream.c:362]: Ah, missed that, you're right.

**Comment by indutny** [comment]
Landed in 145b76b894c6ed6406d7d3d47abedea5b29d942c, thank you!

**Comment by saghul** [comment]
Awesome!

**Comment by indutny** [comment]
Backported to v0.10 in c60d6af

**Pull request closed without merging**